lib/modules/manager/cargo/artifacts.ts

```diff
@@ -33,24 +36,47 @@ async function cargoUpdate(
   await exec(cmd, execOptions);
 }
 
-export async function updateArtifacts({
-  packageFileName,
-  updatedDeps,
-  newPackageFileContent,
-  config,
-}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
-  logger.debug(`cargo.updateArtifacts(${packageFileName})`);
-
-  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+async function cargoUpdatePrecise(
+  manifestPath: string,
+  updatedDeps: Upgrade[],
+  constraint: string | undefined
+): Promise<void> {
+  // First update all dependencies that have been bumped in `Cargo.toml`.
+  const cmds = [
+    'cargo update --config net.git-fetch-with-cli=true' +
+      ` --manifest-path ${quote(manifestPath)} --workspace`,
+  ];
 
-  if (
-    !isLockFileMaintenance &&
-    (updatedDeps === undefined || updatedDeps.length < 1)
-  ) {
-    logger.debug('No updated cargo deps - returning null');
-    return null;
+  // Update individual dependencies to their `newVersion`. Necessary when
+  // using the `update-lockfile` rangeStrategy which doesn't touch Cargo.toml.
+  for (const dep of updatedDeps) {
+    cmds.push(
+      `cargo update --config net.git-fetch-with-cli=true` +
+        ` --manifest-path ${quote(manifestPath)}` +
+        ` --package ${dep.packageName!}@${dep.lockedVersion}` +
+        ` --precise ${dep.newVersion}`
+    );
   }
 
+  const execOptions: ExecOptions = {
+    docker: {},
+    toolConstraints: [{ toolName: 'rust', constraint }],
+  };
+
+  await exec(cmds, execOptions);
+}
+
+export async function updateArtifacts(
+  {
+    packageFileName,
+    updatedDeps,
+    newPackageFileContent,
+    config,
+  }: UpdateArtifact,
+  recursionLimit = 10
+): Promise<UpdateArtifactsResult[] | null> {
```

Comment:
please rename and make private, create a new `updateArtifacts` which called the renamed function, so recursion limit isn't part of the public API.

Reply:
This is exactly how it's done in other managers, i.e. bundler. I'm really getting tired of this back and forth. Please be respectful of my time. I will leave some time for you to do a final review pass. Any other issues you can fix yourself. I'm happy to use this branch for my use case.

Reply:
I think it's best that we close this and someone else can pick it up to completion